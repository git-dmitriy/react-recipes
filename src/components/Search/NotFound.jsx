import { useEffect, useState } from 'react';
import { getRandomMeal } from '../../helpers/api';
import { Meal } from '../meals/Meal';

export const NotFound = ({ target }) => {
  const [randomMeal, setRandomMeal] = useState();
  useEffect(() => {
    getRandomMeal().then((data) => {
      setRandomMeal(data.meals[0]);
    });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <div className='flex flex-col-reverse justify-around items-center my-6 px-2'>
      <ul className='rounded-lg overflow-hidden w-4/6 sm:w-1/2'>
        {randomMeal && <Meal {...randomMeal} />}
      </ul>
      <div className='w-full sm:w-1/2 pl-2 flex flex-col items-center text-center'>
        <h2 className=''>
          <span className='font-bold block'>Nothing found</span> for "{target}"
        </h2>
        <div>Try to cook this </div>
        <div className='w-12 h-1 bg-red-500 rounded mt-2 mb-4' />
      </div>
    </div>
  );
};
