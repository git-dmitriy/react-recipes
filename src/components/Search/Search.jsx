import { useState, useRef } from 'react';
import { BsSearch } from 'react-icons/bs';

export const Search = ({ cb = Function.prototype }) => {
  const [searchQuery, setSearchQuery] = useState('');
  const searchInput = useRef();

  const handleSubmit = () => {
    cb(searchQuery);
  };

  const handleKeyDown = (e) => {
    if (e.key === 'Enter') {
      e.preventDefault();
      handleSubmit();
      setSearchQuery('');
      searchInput.current.blur();
    }
  };

  return (
    <form className='flex' onSubmit={handleSubmit}>
      <div className='relative my-2 mx-auto'>
        <input
          role='searchbox'
          type='text'
          className='focus:outline-none focus:ring-2 focus:ring-red-500 shadow rounded-full w-52 sm:w-60 md:w-72 lg:w-80 xl:w-96 border-0 p-3 pl-9 min-w-full'
          placeholder='Search by name...'
          value={searchQuery}
          ref={searchInput}
          onChange={(e) => {
            setSearchQuery(e.target.value);
          }}
          onKeyDown={handleKeyDown}
        />
        <span
          className='absolute left-0 top-1/2 transform -translate-y-1/2 ml-4 '
          onClick={handleSubmit}
        >
          <BsSearch className='fill-current' />
        </span>
      </div>
    </form>
  );
};


