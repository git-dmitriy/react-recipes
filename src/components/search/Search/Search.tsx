import {
  useState,
  useRef,
  KeyboardEventHandler,
  ChangeEventHandler,
} from 'react';
import { useNavigate } from 'react-router-dom';
import { BsSearch } from 'react-icons/bs';

export const Search: React.FC = () => {
  const [searchQuery, setSearchQuery] = useState('');
  const searchInputRef = useRef<HTMLInputElement | null>(null);
  const navigate = useNavigate();

  const handleSearch = (searchQuery: string) => {
    if (searchQuery.trim().length !== 0) {
      navigate({
        pathname: '/recipes',
        search: `search=${searchQuery}`,
      });
    }
  };

  const onChangeHanlder: ChangeEventHandler<HTMLInputElement> = (event) => {
    event.preventDefault();
    setSearchQuery(event.target.value);
  };

  const onSubmitHanlder = () => {
    handleSearch(searchQuery);
  };

  const onKeyDownHanlder: KeyboardEventHandler = (e) => {
    if (e.key === 'Enter') {
      e.preventDefault();
      onSubmitHanlder();
      setSearchQuery('');
      if (searchInputRef.current !== null) searchInputRef.current.blur();
    }
  };

  return (
    <form className='flex' onSubmit={onSubmitHanlder}>
      <div className='relative my-2 mx-auto dark:text-gray-900'>
        <input
          role='searchbox'
          type='text'
          className='focus:outline-none focus:ring-2 dark:placeholder:text-gray-600 dark:bg-gray-100 focus:ring-red-500 shadow rounded-full w-52 sm:w-60 md:w-72 lg:w-80 xl:w-96 border-0 p-3 pl-9 min-w-full'
          placeholder='Search by name...'
          value={searchQuery}
          ref={searchInputRef}
          onChange={onChangeHanlder}
          onKeyDown={onKeyDownHanlder}
        />
        <span
          className='absolute left-0 top-1/2 transform -translate-y-1/2 ml-4 '
          onClick={onSubmitHanlder}
        >
          <BsSearch className='fill-current' />
        </span>
      </div>
    </form>
  );
};
