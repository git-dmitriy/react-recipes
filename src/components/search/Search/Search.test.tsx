import { Search } from './Search';
import renderer from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter, Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';

it('should render correctly', () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <Search />
      </MemoryRouter>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('should navigate to search=${searchQuery} on click', () => {
  const history = createMemoryHistory();
  const searchQuery = 'beef';

  render(
    <Router location={history.location} navigator={history}>
      <Search />
    </Router>
  );

  const searchInput = screen.getByRole('searchbox');

  fireEvent.input(searchInput, { target: { value: searchQuery } });
  fireEvent.keyDown(searchInput, { key: 'Enter', code: 'Enter', charCode: 13 });

  const { pathname, search } = history.location;
  const path = pathname + search;

  expect(path).toBe(`/recipes?search=${searchQuery}`);
});

it('should stay at the same path on click', () => {
  const history = createMemoryHistory();
  const searchQuery = '';

  render(
    <Router location={history.location} navigator={history}>
      <Search />
    </Router>
  );

  const searchInput = screen.getByRole('searchbox');

  fireEvent.input(searchInput, { target: { value: searchQuery } });
  fireEvent.keyDown(searchInput, { key: 'Enter', code: 'Enter', charCode: 13 });

  const { pathname, search } = history.location;
  const path = pathname + search;

  expect(path).toBe('/');
});
