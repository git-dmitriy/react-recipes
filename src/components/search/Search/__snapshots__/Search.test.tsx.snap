// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<form
  className="flex"
  onSubmit={[Function]}
>
  <div
    className="relative my-2 mx-auto dark:text-gray-900"
  >
    <input
      className="focus:outline-none focus:ring-2 dark:placeholder:text-gray-600 dark:bg-gray-100 focus:ring-red-500 shadow rounded-full w-52 sm:w-60 md:w-72 lg:w-80 xl:w-96 border-0 p-3 pl-9 min-w-full"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Search by name..."
      role="searchbox"
      type="text"
      value=""
    />
    <span
      className="absolute left-0 top-1/2 transform -translate-y-1/2 ml-4 "
      onClick={[Function]}
    >
      <svg
        className="fill-current"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1.007 1.007 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0z"
        />
      </svg>
    </span>
  </div>
</form>
`;
